File: examples/typescript/05-data-structures/main.ts
/*
Data Structures in TypeScript
========================
Demonstrates arrays, tuples, objects, Maps, Sets, and common operations.
TypeScript adds static typing to JavaScript.
*/

// ── Arrays (typed dynamic arrays) ─────────────────

function demonstrateArrays(): void {
  console.log("--- Arrays ---");

  // Typed arrays
  const numbers: number[] = [10, 20, 30, 40, 50];
  console.log("numbers =", numbers);

  // Indexing
  console.log("numbers[0] =", numbers[0]);      // 10
  console.log("numbers[4] =", numbers[4]);      // 50

  // Tuple arrays (heterogeneous)
  const mixed: (string | number)[] = ["hello", 42, "world", 99];
  console.log("mixed =", mixed);

  // Array methods
  numbers.push(60);
  console.log("After push:", numbers);

  numbers.pop();
  console.log("After pop:", numbers);

  // Spread operator
  const moreNumbers = [...numbers, 70, 80];
  console.log("spread:", moreNumbers);

  // Array.from (from iterable)
  const range = Array.from({ length: 5 }, (_, i) => i + 1);
  console.log("range 1-5:", range);

  // map, filter, reduce
  const squares = numbers.map(n => n * n);
  console.log("squares:", squares);

  const evens = numbers.filter(n => n % 2 === 0);
  console.log("evens:", evens);

  const sum = numbers.reduce((acc, n) => acc + n, 0);
  console.log("sum:", sum);
}

// ── Tuples (fixed-size, typed) ─────────────────────

function demonstrateTuples(): void {
  console.log("\n--- Tuples ---");

  // Type annotation for tuple
  const point: [number, number] = [10, 20];
  console.log("point =", point);

  // Destructuring
  const [x, y] = point;
  console.log(`x = ${x}, y = ${y}`);

  // Tuple with more elements
  const rgb: [number, number, number] = [255, 0, 128];
  console.log("RGB =", rgb);

  // Named tuples (objects as tuples)
  type Point3D = [number, number, number];
  const point3d: Point3D = [10, 20, 30];
  console.log("3D point =", point3d);

  const [, , z] = point3d;  // Ignore first two, get third
  console.log("z =", z);
}

// ── Objects (key-value mappings) ─────────────────────

function demonstrateObjects(): void {
  console.log("\n--- Objects ---");

  // Creation
  const person = {
    name: "Ada",
    age: 36,
    city: "London"
  };
  console.log("person =", person);

  // Access
  console.log("name =", person.name);
  console.log("age =", person["age"]);  // Bracket notation

  // Optional chaining (non-null assertion)
  console.log("email =", person.email ?? "N/A");

  // Modification
  person.email = "ada@example.com";
  console.log("After adding email:", person);

  delete person.city;
  console.log("After deleting city:", person);

  // Object spread (cloning)
  const person2 = { ...person, country: "UK" };
  console.log("cloned person =", person2);

  // Object keys/values
  console.log("Keys:", Object.keys(person));
  console.log("Values:", Object.values(person));
  console.log("Entries:", Object.entries(person));

  // Object.freeze (immutable object)
  const frozen = Object.freeze({ x: 10, y: 20 });
  // frozen.x = 30;  // Error in strict mode
  console.log("frozen =", frozen);
}

// ── Map (key-value with any key type) ───────────

function demonstrateMap(): void {
  console.log("\n--- Map ---");

  // Creation
  const scores = new Map<string, number>();
  scores.set("Alice", 95);
  scores.set("Bob", 87);
  scores.set("Charlie", 92);

  console.log("scores =", Array.from(scores.entries()));

  // Access
  console.log("Alice's score =", scores.get("Alice"));

  // Check existence
  console.log("has 'Bob'? =", scores.has("Bob"));

  // Delete
  scores.delete("Bob");
  console.log("After delete Bob:", Array.from(scores.entries()));

  // Size
  console.log("size =", scores.size);

  // Iteration
  console.log("Iteration:");
  scores.forEach((score, name) => {
    console.log(`  ${name} -> ${score}`);
  });

  // WeakMap (garbage collectable)
  const weak = new WeakMap<object, string>();
  const key = {};
  weak.set(key, "value");
  console.log("WeakMap has key?", weak.has(key));
}

// ── Set (unique values) ─────────────────────────────

function demonstrateSet(): void {
  console.log("\n--- Set ---");

  // Creation (duplicates removed)
  const tags = new Set<string>(["python", "typing", "python", "data"]);
  console.log("tags =", Array.from(tags));
  console.log("size =", tags.size);  // 3

  // Methods
  console.log("has 'typing'? =", tags.has("typing"));

  tags.add("algorithms");
  console.log("After add 'algorithms':", Array.from(tags));

  tags.delete("data");
  console.log("After delete 'data':", Array.from(tags));

  // Set operations
  const setA = new Set<number>([1, 2, 3, 4, 5]);
  const setB = new Set<number>([4, 5, 6, 7, 8]);

  const union = new Set([...setA, ...setB]);
  console.log("union =", Array.from(union));

  const intersection = new Set([...setA].filter(x => setB.has(x)));
  console.log("intersection =", Array.from(intersection));

  const difference = new Set([...setA].filter(x => !setB.has(x)));
  console.log("difference A-B =", Array.from(difference));
}

// ── Interfaces (type contracts) ─────────────────────

interface Point {
  x: number;
  y: number;
}

interface Person {
  name: string;
  age: number;
}

function demonstrateInterfaces(): void {
  console.log("\n--- Interfaces ---");

  const point: Point = { x: 10, y: 20 };
  console.log("Point:", point);

  const ada: Person = { name: "Ada", age: 36 };
  console.log("Person:", ada);

  // Type guards
  function isPoint(value: unknown): value is Point {
    return typeof value === "object" && "x" in value && "y" in value;
  }

  const value = { x: 10, y: 20 };
  if (isPoint(value)) {
    console.log("value is a Point:", value);
  }
}

// ── Type Aliases ─────────────────────────────────────

type ID = string | number;
type Coordinates = [number, number];

function demonstrateTypeAliases(): void {
  console.log("\n--- Type Aliases ---");

  const userId: ID = 123;
  console.log("userId =", userId);  // string | number

  const coords: Coordinates = [10, 20];
  console.log("coords =", coords);
}

// ── Union Types ─────────────────────────────────────

function demonstrateUnionTypes(): void {
  console.log("\n--- Union Types ---");

  type Result = number | string;

  function process(value: Result): void {
    if (typeof value === "number") {
      console.log("Number result:", value * 2);
    } else {
      console.log("String result:", value.toUpperCase());
    }
  }

  process(42);
  process("hello");
}

// ── Optional chaining ─────────────────────────────────

function demonstrateOptional(): void {
  console.log("\n--- Optional chaining ---");

  interface User {
    name: string;
    email?: string;  // Optional
    phone?: string;  // Optional
  }

  const user: User = {
    name: "Ada"
  };

  console.log("name:", user.name);
  console.log("email (with default):", user.email ?? "N/A");
  console.log("phone (with default):", user.phone ?? "N/A");
}

// ── Readonly (immutable) ─────────────────────────────

function demonstrateReadonly(): void {
  console.log("\n--- Readonly ---");

  const mutable = [1, 2, 3];
  const readonly: ReadonlyArray<number> = [1, 2, 3] as readonly;

  console.log("mutable:", mutable);
  // mutable.push(4);  // OK

  console.log("readonly:", readonly);
  // readonly.push(4);  // Error!

  // Readonly object
  const config = { port: 8080, host: "localhost" } as const;
  // config.port = 9090;  // Error!
}

// ── Summary ───────────────────────────────────────────

function main(): void {
  console.log("Data Structures in TypeScript");
  console.log("==============================");

  demonstrateArrays();
  demonstrateTuples();
  demonstrateObjects();
  demonstrateMap();
  demonstrateSet();
  demonstrateInterfaces();
  demonstrateTypeAliases();
  demonstrateUnionTypes();
  demonstrateOptional();
  demonstrateReadonly();

  console.log("\n--- Summary ---");
  console.log("TypeScript data structures:");
  console.log("  - Arrays: typed, with spread operator");
  console.log("  - Tuples: fixed-size, typed");
  console.log("  - Objects: key-value with optional chaining");
  console.log("  - Map: any key type, ordered insertion");
  console.log("  - Set: unique values, with operations");
  console.log("  - Interfaces: type contracts and duck typing");
  console.log("  - Type aliases: named types");
  console.log("  - Union types: multiple possible types");
  console.log("  - Optional: nullable with ?");
  console.log("  - Readonly: immutable with const/as readonly");
}

File: examples/haskell/05-data-structures/Main.hs
-- Data Structures in Haskell
-- =========================
-- Demonstrates lists, Data.Map, Data.Set, custom data types.
-- Haskell provides persistent, lazy data structures with type safety.

module Main where

import qualified Data.Map as Map
import qualified Data.Set as Set

-- ── Lists (linked, lazy) ─────────────────────

demonstrateLists :: IO ()
demonstrateLists = do
  putStrLn "--- Lists (linked, lazy) ---"

  -- List literal
  let fruits = ["apple", "banana", "cherry"]
  putStrLn $ "fruits = " ++ show fruits

  -- Indexing (head, tail, !! operator)
  putStrLn $ "head = " ++ show (head fruits)        -- apple
  putStrLn $ "tail = " ++ show (tail fruits)        -- ["banana", "cherry"]
  putStrLn $ "fruits !! 0 = " ++ show (fruits !! 0)  -- apple

  -- Pattern matching (destructuring)
  let (first:rest) = fruits
  putStrLn $ "first = " ++ show first
  putStrLn $ "rest = " ++ show rest

  -- Cons (prepend)
  let fruits2 = "date" : fruits
  putStrLn $ "After cons 'date': " ++ show fruits2

  -- List comprehension
  let squares = [x * x | x <- [1..10]]
  putStrLn $ "squares (1-10): " ++ show squares

  let evens = [x | x <- [1..20], even x]
  putStrLn $ "evens (1-20): " ++ show evens

  -- map, filter, fold
  let doubled = map (*2) [1..5]
  putStrLn $ "doubled (1-5): " ++ show doubled

  let sumList = foldl (+) 0 [1..10]
  putStrLn $ "sum (1-10): " ++ show sumList

  let product = foldl (*) 1 [1..5]
  putStrLn $ "product (1-5): " ++ show product

-- ── Tuples (fixed-size, heterogenous) ─────────────

demonstrateTuples :: IO ()
demonstrateTuples = do
  putStrLn "\n--- Tuples (fixed-size) ---"

  -- Pair
  let point = (10, 20)
  putStrLn $ "point = " ++ show point

  -- Destructuring
  let (x, y) = point
  putStrLn $ "x = " ++ show x ++ ", y = " ++ show y

  -- Triple
  let rgb = (255, 0, 128)
  putStrLn $ "RGB = " ++ show rgb

  -- Nested tuples
  let nested = ((1, 2), (3, 4))
  putStrLn $ "nested = " ++ show nested

-- ── Maps (key-value) ─────────────────────────────

demonstrateMaps :: IO ()
demonstrateMaps = do
  putStrLn "\n--- Maps (key-value) ---"

  -- Create map from list
  let person = Map.fromList [("name", "Ada"), ("age", 36), ("city", "London")]
  putStrLn $ "person = " ++ show (Map.toList person)

  -- Access (Maybe for missing keys)
  putStrLn $ "name = " ++ show (Map.lookup "name" person)
  putStrLn $ "age = " ++ show (Map.lookup "age" person)
  putStrLn $ "email = " ++ show (Map.lookup "email" person)

  -- Insert
  let person2 = Map.insert "email" "ada@example.com" person
  putStrLn $ "After insert email: " ++ show (Map.toList person2)

  -- Delete
  let person3 = Map.delete "city" person2
  putStrLn $ "After delete city: " ++ show (Map.toList person3)

  -- Check membership
  putStrLn $ "has 'email'? " ++ show (Map.member "email" person3)

-- ── Sets (unique elements) ─────────────────────

demonstrateSets :: IO ()
demonstrateSets = do
  putStrLn "\n--- Sets (unique elements) ---"

  -- Create set (duplicates removed)
  let tags = Set.fromList ["python", "typing", "python", "data"]
  putStrLn $ "tags = " ++ show (Set.toList tags)
  putStrLn $ "size = " ++ show (Set.size tags)

  -- Operations
  let setA = Set.fromList [1..5]
  let setB = Set.fromList [4..8]

  let union = Set.union setA setB
  putStrLn $ "union: " ++ show (Set.toList union)

  let intersection = Set.intersection setA setB
  putStrLn $ "intersection: " ++ show (Set.toList intersection)

  let difference = Set.difference setA setB
  putStrLn $ "difference A-B: " ++ show (Set.toList difference)

  -- Subset check
  let subset = Set.fromList [2, 3, 4]
  putStrLn $ "is [2,3,4] subset of [1..5]? " ++ show (Set.isSubsetOf subset setA)

-- ── Custom Data Types ─────────────────────────────

data Point = Point { x :: Int, y :: Int }

data Person = Person { name :: String, age :: Int }

data Status = Active | Inactive | Pending String

demonstrateCustomTypes :: IO ()
demonstrateCustomTypes = do
  putStrLn "\n--- Custom Data Types ---"

  -- Record syntax for constructing
  let p = Point 10 20
  putStrLn $ "Point = " ++ show p

  -- Field access
  putStrLn $ "x = " ++ show (x p)
  putStrLn $ "y = " ++ show (y p)

  -- Person with string field
  let ada = Person "Ada" 36
  putStrLn $ "Person = " ++ show ada

  -- Enum variants
  let status1 = Active
  let status2 = Inactive
  let status3 = Pending "Waiting for approval"

  putStrLn $ "status1 = " ++ show status1
  putStrLn $ "status2 = " ++ show status2
  putStrLn $ "status3 = " ++ show status3

  -- Pattern matching on enums
  case status3 of
    Pending msg -> putStrLn $ "Pending: " ++ msg
    _ -> putStrLn $ "Not pending"

-- ── Lazy Evaluation ─────────────────────────────────

demonstrateLazy :: IO ()
demonstrateLazy = do
  putStrLn "\n--- Lazy Evaluation ---"

  -- Infinite list
  let naturals = [1..]
  putStrLn $ "First 5 of infinite: " ++ show (take 5 naturals)

  -- Lazy map
  let squares = map (^2) [1..]
  putStrLn $ "First 5 of lazy squares: " ++ show (take 5 squares)

  -- Lazy filter
  let evens = filter even [1..]
  putStrLn $ "First 5 of lazy evens: " ++ show (take 5 evens)

  putStrLn "Lazy sequences are computed on demand!"

-- ── Newtype (type wrapper) ─────────────────────

newtype UserId = UserId String

demonstrateNewtype :: IO ()
demonstrateNewtype = do
  putStrLn "\n--- Newtype (type wrapper) ---"

  let uid = UserId "user123"
  putStrLn $ "UserId = " ++ show uid

  -- Unwrap (pattern matching)
  case uid of
    UserId id -> putStrLn $ "id = " ++ id

-- ── Maybe (optional values) ─────────────────────────

demonstrateMaybe :: IO ()
demonstrateMaybe = do
  putStrLn "\n--- Maybe (optional values) ---"

  -- Just value
  let maybe1 = Just 42
  putStrLn $ "maybe1 = " ++ show maybe1

  -- Nothing (no value)
  let maybe2 = Nothing
  putStrLn $ "maybe2 = " ++ show maybe2

  -- Pattern matching
  case maybe1 of
    Just n -> putStrLn $ "Just contains: " ++ show n
    Nothing -> putStrLn $ "Nothing, no value"

  -- Maybe monadic operations
  let safeDivide _ 0 = Nothing
  let safeDivide x y = Just (x `div` y)

  putStrLn $ "10 / 0 = " ++ show (safeDivide 10 0)
  putStrLn $ "10 / 3 = " ++ show (safeDivide 10 3)

-- ── Summary ───────────────────────────────────────────

main :: IO ()
main = do
  putStrLn "Data Structures in Haskell"
  putStrLn "========================"

  demonstrateLists()
  demonstrateTuples()
  demonstrateMaps()
  demonstrateSets()
  demonstrateCustomTypes()
  demonstrateLazy()
  demonstrateNewtype()
  demonstrateMaybe()

  putStrLn "\n--- Summary ---"
  putStrLn "Haskell data structures:"
  putStrLn "  - List: linked, lazy, cons for prepend"
  putStrLn "  - Map: Data.Map (ordered balanced tree)"
  putStrLn "  - Set: Data.Set (ordered balanced tree)"
  putStrLn "  - Tuples: fixed-size heterogenous"
  putStrLn "  - Algebraic data types: sum types with pattern matching"
  putStrLn "  - Records: field access with dot syntax"
  putStrLn "  - Maybe: optional values (no null pointer)"
  putStrLn "  - Lazy evaluation: computed on demand"
  putStrLn "  - Type classes: ad-hoc polymorphism"
  putStrLn "  - Newtype: zero-overhead wrapper"
